src/views/Users/ViewUser/index.jsx

```diff
@@ -0,0 +1,394 @@
+  supportsActionRestriction,
+  getSupportedActions,
+} from '../../../utils/Users';
+
+const useStyles = makeStyles(theme => ({
+  fab: {
+    ...theme.mixins.fab,
+    right: theme.spacing(12),
+  },
+  fullWidth: {
+    width: '100%',
+  },
+  addGrid: {
+    marginTop: theme.spacing(0),
+  },
+  gridWithIcon: {
+    marginTop: theme.spacing(3),
+  },
+  gridDelete: {
+    display: 'flex',
+    justifyContent: 'flex-end',
+  },
+}));
+
+function ViewUser({ isNewUser, ...props }) {
+  const getEmptyPermission = () => ({
+    name: '',
+    options: {
+      products: [],
+      actions: [],
+    },
+    metadata: {
+      isAdditional: true,
+      productText: '',
+      actionText: '',
+    },
+  });
+  const getEmptyRole = () => ({
+    name: '',
+    data_version: null,
+    metadata: {
+      isAdditional: true,
+    },
+  });
+  const {
+    match: {
+      params: { username: existingUsername },
+    },
+  } = props;
+  const classes = useStyles();
+  const [username, setUsername] = useState('');
+  const [roles, setRoles] = useState([]);
+  const [originalRoles, setOriginalRoles] = useState([]);
+  const [additionalRoles, setAdditionalRoles] = useState([]);
+  const [permissions, setPermissions] = useState([]);
+  const [originalPermissions, setOriginalPermissions] = useState([]);
+  const [additionalPermissions, setAdditionalPermissions] = useState(
+    isNewUser ? [getEmptyPermission()] : []
+  );
+  const [products, setProducts] = useState([]);
+  const [productsAction, fetchProducts] = useAction(getProducts);
+  const [userAction, fetchUser] = useAction(getUserInfo);
+  const [saveAction, saveUser] = useAction(() => {});
+  const isLoading = userAction.loading || productsAction.loading;
+  const error = userAction.error || productsAction.error || saveAction.error;
+  const defaultToEmptyString = defaultTo('');
+
+  useEffect(() => {
+    if (!isNewUser) {
+      Promise.all([fetchUser(existingUsername), fetchProducts()]).then(
+        ([userdata, productdata]) => {
+          const roles = Object.keys(userdata.data.data.roles).map(name => ({
+            name,
+            data_version: userdata.data.data.roles[name].data_version,
+            metadata: {
+              isAdditional: false,
+            },
+          }));
+          const permissions = Object.keys(userdata.data.data.permissions).map(
+            name => {
+              const details = userdata.data.data.permissions[name];
+
+              return {
+                name,
+                options: details.options || { products: [], actions: [] },
+                data_version: details.data_version,
+                metadata: {
+                  isAdditional: false,
+                  productText: '',
+                  actionText: '',
+                },
+              };
+            }
+          );
+
+          setUsername(userdata.data.data.username);
+          setRoles(roles);
+          setOriginalRoles(JSON.parse(JSON.stringify(roles)));
```

Comment:
The reason we're doing it for originalRoles is to ensure we have a separate instance of the object (so that when someone removes an existing role, it doesn't also change originalRoles). So I don't think we need to do the same for roles.

Reply:
Oh I see. To make it more explicit, we could use the `clone` method from ramda:

```jsx
import { clone } from 'ramda';

// ...

setOriginalRoles(clone(roles));
```